psl-core/src/main/java/org/linqs/psl/reasoner/dcd/DCDReasoner.java

```diff
@@ -59,82 +60,89 @@ public double optimize(TermStore baseTermStore) {
+        double objective = Double.POSITIVE_INFINITY;
         double oldObjective = Double.POSITIVE_INFINITY;
+        float[] oldVariableValues = null;
 
         if (printInitialObj && log.isTraceEnabled()) {
-            objective = computeObjective(termStore, variableValues);
-            log.trace("Iteration {} -- Objective: {}, Iteration Time: {}, Total Optimiztion Time: {}", 0, objective, 0, 0);
+            objective = computeObjective(termStore);
+            log.trace("Iteration {} -- Objective: {}, Mean Movement: {}, Iteration Time: {}, Total Optimiztion Time: {}", 0, objective, 0.0f, 0, 0);
         }
 
-        int iteration = 1;
         long totalTime = 0;
-        while (true) {
+        boolean converged = false;
+        for (int iteration = 1; iteration < (maxIterations * budget) && !converged; iteration++) {
             long start = System.currentTimeMillis();
 
+            termCount = 0;
+            objective = 0;
             for (DCDObjectiveTerm term : termStore) {
-                term.minimize(truncateEveryStep, variableValues);
+                if (oldVariableValues != null) {
+                    objective += term.evaluate(oldVariableValues);
+                } else {
+                    objective += term.evaluate(termStore.getVariableValues());
+                }
+
+                termCount++;
+                term.minimize(truncateEveryStep, termStore.getVariableValues());
             }
 
+            termStore.iterationComplete();
+
             // If we are truncating every step, then the variables are already in valid state.
             if (!truncateEveryStep) {
-                for (RandomVariableAtom variable : termStore.getVariables()) {
-                    variable.setValue(Math.max(Math.min(variable.getValue(), 1.0f), 0.0f));
+                for (int i = 0; i < termStore.getNumVariables(); i++) {
+                    termStore.getVariableValues()[i] = Math.max(0.0f, Math.min(1.0f, termStore.getVariableValues()[i]));
                 }
             }
 
-            long end = System.currentTimeMillis();
-
-            oldObjective = objective;
-            objective = computeObjective(termStore, variableValues);
-            totalTime += end - start;
+            converged = breakOptimization(objective, oldObjective, termCount);
 
-            if (log.isTraceEnabled()) {
-                log.trace("Iteration {} -- Objective: {}, Iteration Time: {}, Total Optimiztion Time: {}",
-                        iteration, objective, (end - start), totalTime);
+            // Keep track of the old variables for a deferred objective computation.
+            if (oldVariableValues == null) {
+                oldVariableValues = Arrays.copyOf(termStore.getVariableValues(), termStore.getVariableValues().length);
+                oldObjective = Double.POSITIVE_INFINITY;
+            } else {
+                System.arraycopy(termStore.getVariableValues(), 0, oldVariableValues, 0, oldVariableValues.length);
```

Comment:
What is oldVariableValues used for? Why do we need to make a copy?

Reply:
Old variable values are used for a delayed objective calculation that is performed as the terms are being minimized. This saves passes through the data which becomes important for large-scale reasoning and streaming inference when passes through the data require reading terms off disk.